Diagnostic for forgetting a writer in `write!` is bad

### Code

```Rust
fn main() {
    let x = 1;
    let y = 2;
    write!("{}_{}", x, y);
}
```


### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error: format argument must be a string literal
 --> src/main.rs:4:21
  |
4 |     write!("{}_{}", x, y);
  |                     ^
  |
help: you might be missing a string literal to format with
  |
4 |     write!("{}_{}", "{} {}", x, y);
  |                     ++++++++

error[E0599]: no method named `write_fmt` found for reference `&'static str` in the current scope
 --> src/main.rs:4:5
  |
4 |     write!("{}_{}", x, y);
  |     ^^^^^^^^^^^^^^^^^^^^^ method not found in `&str`
  |
  = note: this error originates in the macro `write` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to 2 previous errors
```


### Desired output

```Shell
error: cannot write into `&'static str`
 --> src/main.rs:4:21
  |
4 |     write!("{}_{}", x, y);
  |             ^^^^
  |
help: you might be missing a writer to write into
  +    let mut writer = String::new();
  |
4 |     write!(&mut writer, "{}_{}", x, y);
  |            ++++++++++++
```


### Rationale and extra context

_No response_

### Other cases

_No response_

### Anything else?

_No response_

Got it, thank you so much for pointing me in the right direction 🙂

As far as the write_fmt, this is where I am so far: 

```
        let is_write_macro = sugg_span.ctxt().outer_expn_data().kind
            == ExpnKind::Macro(MacroKind::Bang, Symbol::intern("write"));
        let mut err;
        let is_write_fmt = item_name.name == Symbol::intern("write_fmt");
        if is_write_fmt && is_write_macro && args.is_some() {
            let helper = tcx.sess.source_map().span_to_snippet(args.unwrap().0.span).unwrap();
            let writer_note;
            match args.unwrap().0.kind {
                ExprKind::Lit(_) => {
                    writer_note = Some("you might want to insert a writer in front of this format string")
                }
                _ => writer_note = None,
            };
            err = struct_span_err!(
                tcx.sess,
                args.unwrap().0.span,
                E0599,
                "cannot write into '{}'",
                ty_str_reported,
            );
            err.note(format!(
                "{} must implement 'io::Write', 'fmt::Write' or have a 'write_fmt' method",
                helper
            ));
            if writer_note.is_some() {
                err.help(writer_note.unwrap());
            }
        }
```

Which gives the following output:
```
error[E0599]: cannot write into '&'static str'
 --> src/macros/mod.rs:484:12
  |
6 |     write!("{}_{}", x, y);
  |     -------^^^^^^^------- method not found in `&str`
  |
  = note: "{}_{}" must implement 'io::Write', 'fmt::Write' or have a 'write_fmt' method
  = help: you might want to insert a writer in front of this format string
```

Where I am checking to see if the function is write_fmt and the macro is write! and then creating the relevant error message.

I still need to create the recommended substitution from your example, I think that would be from rustc_errors::CodeSuggestions, is that correct? 

Then for the format.rs error about the missing format string, using ecx.expansion_cause() gives me the whole line of the write! macro `write!("{}_{}", x, y)`, I can think of some hacky ways to pull write! out of the outer span, is there a better way to extract the outer macro or should I just focus on improving the suggest.rs portion of the diagnostic?

Thanks again!
James

--- a/compiler/rustc_hir_typeck/src/method/suggest.rs
+++ b/compiler/rustc_hir_typeck/src/method/suggest.rs
@@ -245,6 +245,28 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         None
     }
 
+    fn suggest_missing_writer(
+        &self,
+        rcvr_ty: Ty<'tcx>,
+        args: (&'tcx hir::Expr<'tcx>, &'tcx [hir::Expr<'tcx>]),
+    ) -> DiagnosticBuilder<'_, ErrorGuaranteed> {
+        let (ty_str, _ty_file) = self.tcx.short_ty_string(rcvr_ty);
+        let mut err =
+            struct_span_err!(self.tcx.sess, args.0.span, E0599, "cannot write into `{}`", ty_str);
+        err.span_note(
+            args.0.span,
+            "must implement `io::Write`, `fmt::Write`, or have a `write_fmt` method",
+        );
+        if let ExprKind::Lit(_) = args.0.kind {
+            err.span_help(
+                args.0.span.shrink_to_lo(),
+                "a writer is needed before this format string",
+            );
+        };
+
+        err
+    }
+
     pub fn report_no_match_method_error(
         &self,
         mut span: Span,
@@ -323,16 +345,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
         }
 
-        let mut err = struct_span_err!(
-            tcx.sess,
-            span,
-            E0599,
-            "no {} named `{}` found for {} `{}` in the current scope",
-            item_kind,
-            item_name,
-            rcvr_ty.prefix_string(self.tcx),
-            ty_str_reported,
-        );
+        let is_write = sugg_span.ctxt().outer_expn_data().macro_def_id.map_or(false, |def_id| {
+            tcx.is_diagnostic_item(sym::write_macro, def_id)
+                || tcx.is_diagnostic_item(sym::writeln_macro, def_id)
+        }) && item_name.name == Symbol::intern("write_fmt");
+        let mut err = if is_write
+            && let Some(args) = args
+        {
+            self.suggest_missing_writer(rcvr_ty, args)
+        } else {
+            struct_span_err!(
+                tcx.sess,
+                span,
+                E0599,
+                "no {} named `{}` found for {} `{}` in the current scope",
+                item_kind,
+                item_name,
+                rcvr_ty.prefix_string(self.tcx),
+                ty_str_reported,
+            )
+        };
         if tcx.sess.source_map().is_multiline(sugg_span) {
             err.span_label(sugg_span.with_hi(span.lo()), "");
         }

--- a/tests/ui/macros/missing-writer.rs
+++ b/tests/ui/macros/missing-writer.rs
@@ -1,0 +1,17 @@
+// Check error for missing writer in writeln! and write! macro
+fn main() {
+    let x = 1;
+    let y = 2;
+    write!("{}_{}", x, y);
+    //~^ ERROR format argument must be a string literal
+    //~| HELP you might be missing a string literal to format with
+    //~| ERROR cannot write into `&'static str`
+    //~| NOTE must implement `io::Write`, `fmt::Write`, or have a `write_fmt` method
+    //~| HELP a writer is needed before this format string
+    writeln!("{}_{}", x, y);
+    //~^ ERROR format argument must be a string literal
+    //~| HELP you might be missing a string literal to format with
+    //~| ERROR cannot write into `&'static str`
+    //~| NOTE must implement `io::Write`, `fmt::Write`, or have a `write_fmt` method
+    //~| HELP a writer is needed before this format string
+}

--- a/tests/ui/macros/missing-writer.stderr
+++ b/tests/ui/macros/missing-writer.stderr
@@ -1,0 +1,59 @@
+error: format argument must be a string literal
+  --> $DIR/missing-writer.rs:5:21
+   |
+LL |     write!("{}_{}", x, y);
+   |                     ^
+   |
+help: you might be missing a string literal to format with
+   |
+LL |     write!("{}_{}", "{} {}", x, y);
+   |                     ++++++++
+
+error: format argument must be a string literal
+  --> $DIR/missing-writer.rs:11:23
+   |
+LL |     writeln!("{}_{}", x, y);
+   |                       ^
+   |
+help: you might be missing a string literal to format with
+   |
+LL |     writeln!("{}_{}", "{} {}", x, y);
+   |                       ++++++++
+
+error[E0599]: cannot write into `&'static str`
+  --> $DIR/missing-writer.rs:5:12
+   |
+LL |     write!("{}_{}", x, y);
+   |     -------^^^^^^^------- method not found in `&str`
+   |
+note: must implement `io::Write`, `fmt::Write`, or have a `write_fmt` method
+  --> $DIR/missing-writer.rs:5:12
+   |
+LL |     write!("{}_{}", x, y);
+   |            ^^^^^^^
+help: a writer is needed before this format string
+  --> $DIR/missing-writer.rs:5:12
+   |
+LL |     write!("{}_{}", x, y);
+   |            ^
+
+error[E0599]: cannot write into `&'static str`
+  --> $DIR/missing-writer.rs:11:14
+   |
+LL |     writeln!("{}_{}", x, y);
+   |     ---------^^^^^^^------- method not found in `&str`
+   |
+note: must implement `io::Write`, `fmt::Write`, or have a `write_fmt` method
+  --> $DIR/missing-writer.rs:11:14
+   |
+LL |     writeln!("{}_{}", x, y);
+   |              ^^^^^^^
+help: a writer is needed before this format string
+  --> $DIR/missing-writer.rs:11:14
+   |
+LL |     writeln!("{}_{}", x, y);
+   |              ^
+
+error: aborting due to 4 previous errors
+
+For more information about this error, try `rustc --explain E0599`.

--- a/tests/ui/suggestions/mut-borrow-needed-by-trait.stderr
+++ b/tests/ui/suggestions/mut-borrow-needed-by-trait.stderr
@@ -21,18 +21,22 @@ note: required by a bound in `BufWriter`
   --> $SRC_DIR/std/src/io/buffered/bufwriter.rs:LL:COL
 
 error[E0599]: the method `write_fmt` exists for struct `BufWriter<&dyn Write>`, but its trait bounds were not satisfied
-  --> $DIR/mut-borrow-needed-by-trait.rs:21:5
+  --> $DIR/mut-borrow-needed-by-trait.rs:21:14
    |
 LL |     writeln!(fp, "hello world").unwrap();
-   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ method cannot be called on `BufWriter<&dyn Write>` due to unsatisfied trait bounds
+   |     ---------^^---------------- method cannot be called on `BufWriter<&dyn Write>` due to unsatisfied trait bounds
   --> $SRC_DIR/std/src/io/buffered/bufwriter.rs:LL:COL
    |
    = note: doesn't satisfy `BufWriter<&dyn std::io::Write>: std::io::Write`
    |
+note: must implement `io::Write`, `fmt::Write`, or have a `write_fmt` method
+  --> $DIR/mut-borrow-needed-by-trait.rs:21:14
+   |
+LL |     writeln!(fp, "hello world").unwrap();
+   |              ^^
    = note: the following trait bounds were not satisfied:
            `&dyn std::io::Write: std::io::Write`
            which is required by `BufWriter<&dyn std::io::Write>: std::io::Write`
-   = note: this error originates in the macro `writeln` (in Nightly builds, run with -Z macro-backtrace for more info)
 
 error: aborting due to 3 previous errors
 

